### ISSUE TITLE
mix deps.get fails frequently

### ISSUE THREAD

**Issue by astery** [hexpm/hex#308]
Just create any mix project with any dependencies, run multiple times `mix deps.get`. If there are many dependencies error became to appear more often. In my project I can't reach complete phase.

If this is hex server fail, can we get ability to complete after fail? And also more descriptive error?

```
$ mix deps.get
** (exit) exited in: GenServer.call(#PID<0.114.0>, {:versions, "db_connection"}, 30000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:194: Hex.RemoteConverger.check_dep/2
    (elixir) lib/enum.ex:657: anonymous fn/3 in Enum.each/2
    (stdlib) lists.erl:1263: :lists.foldl/3
    (elixir) lib/enum.ex:656: Enum.each/2
    (hex) lib/hex/remote_converger.ex:39: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
```

```
$ mix hex.info
Hex:    0.14.0
Elixir: 1.3.3
OTP:    19.0.3

Built with: Elixir 1.3.4 and OTP 18.3.4.4
```

**Comment by astery** [comment]
```
elixir -e ":inets_trace.enable(:max, 'trace.txt', :httpc)" -S mix deps.get
```
[trace.txt](https://github.com/hexpm/hex/files/569758/trace.txt)

**Comment by ericmj** [comment]
Thanks for the report. I will investigate what's going on.

This happens frequently but not 100% of the time, correct?

**Comment by ericmj** [comment]
I have not been able to reproduce this. Can you gist the `mix.exs` and `mix.lock` files for the project?

**Comment by astery** [comment]
This happens often on project with one dependency and 100% on project with 20 > dependencies.

https://gist.github.com/astery/7ffd44de9afb1c2dc2301843f7f75b1f

**Comment by josevalim** [comment]
@ericmj are we trying to download all deps in parallel? Maybe we should do it in chunks? Or using Task.async_stream if available?

**Comment by ericmj** [comment]
@astery How long does it take to run `mix deps.get` on the successful runs?

@josevalim httpc limits the concurrency. I don't want to use `Task.async_stream` because that would only solve it for some users.

**Comment by astery** [comment]
Here is small video
![mix_deps](https://cloud.githubusercontent.com/assets/1290275/20054595/0ce53f86-a4f7-11e6-8c09-75fb992b3ae4.gif)

**Comment by ericmj** [comment]
@astery It seems like the process closed before it could finish writing to the trace file. I assume you also filtered away the `{cacerts, ...}` output from trace.txt? You can run with `HEX_UNSAFE_HTTPS=1`, that should make the tracing quicker, if you'd like to indulge me some more I can provide a build that adds a long sleep to give the tracing time to finish.

**Comment by astery** [comment]
Yes, I filtered `{cacerts, ...}`, forgot to mention it. Here is trace.


`HEX_UNSAFE_HTTPS=1 elixir -e ":inets_trace.enable(:max, 'trace.txt', :httpc)" -S mix deps.get`
[trace.txt](https://github.com/hexpm/hex/files/578346/trace.txt)

**Comment by ericmj** [comment]
My current theory is this happens due to some buggy pipelining in httpc.

@astery Let me know if you can try build with some fixes and where pipelining is disabled.

**Comment by rhruiz** [comment]
Same thing happens to me with 29 deps in `mix.exs`

**Comment by ericmj** [comment]
We appreciate any help we can get investigating this since no one in the core team can reproduce this.

On the `master` branch we have some fixes where could do accidentally do duplicate requests that should improve performance but is not likely the issue you are seeing. There is also an `emj-disable-pipeline` branch where we disable pipelining in the http client. You can install a development version of hex like this:

```
git clone https://github.com/hexpm/hex.git
cd hex
git checkout BRANCH
mix install
```

Change `BRANCH` to `master` or `emj-disable-pipeline` depending on which branch you are testing. You can run `mix local.hex` to install the stable version again.

Thank you for your help!

**Comment by rhruiz** [comment]
@ericmj success on the `emj-disable-pipeline` branch \0/

**Comment by ericmj** [comment]
@rhruiz Thanks for the quick check. Please use the branch for a while and let us know if the issue comes back and please check if it comes back if you install stable again with `mix local.hex`.

**Comment by rhruiz** [comment]
I don't know if this helps or if it is related, but my build process always forces the install of mix and rebar before fetching the deps. And when the problem with the deps happens, installing rebar will usually print the following errors:

```23:02:32.590 [error] Failed updating session: 
   ProfileName: :httpc_mix
   SessionId:   {{'repo.hex.pm', 443}, #PID<0.111.0>}
   Pos:         7
   Value:       0
when
   Session (db) info: :undefined
   Session (db):      {:session, {{'repo.hex.pm', 443}, #PID<0.111.0>}, false, :https,
 {:sslsocket, {:gen_tcp, #Port<0.4237>, :tls_connection, :undefined},
  #PID<0.115.0>}, {:essl, []}, 1, :keep_alive, true}
   Session (record):  {:EXIT,
 {:badarg,
  [{:ets, :lookup,
    [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.111.0>}], []},
   {:httpc_manager, :lookup_session, 2, [file: 'httpc_manager.erl', line: 189]},
   {:httpc_handler, :update_session, 4,
    [file: 'httpc_handler.erl', line: 1909]},
   {:httpc_handler, :handle_empty_queue, 4,
    [file: 'httpc_handler.erl', line: 1368]},
   {:httpc_handler, :handle_info, 2, [file: 'httpc_handler.erl', line: 471]},
   {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]},
   {:gen_server, :handle_msg, 5, [file: 'gen_server.erl', line: 667]},
   {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}]}}
   T: :error
   E: :badarg

23:02:32.591 [error] Failed updating session: 
   ProfileName: :httpc_mix
   SessionId:   {{'repo.hex.pm', 443}, #PID<0.111.0>}
   Pos:         9
   Value:       true
when
   Session (db) info: :undefined
   Session (db):      {:session, {{'repo.hex.pm', 443}, #PID<0.111.0>}, false, :https,
 {:sslsocket, {:gen_tcp, #Port<0.4237>, :tls_connection, :undefined},
  #PID<0.115.0>}, {:essl, []}, 1, :keep_alive, false}
   Session (record):  {:EXIT,
 {:badarg,
  [{:ets, :lookup,
    [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.111.0>}], []},
   {:httpc_manager, :lookup_session, 2, [file: 'httpc_manager.erl', line: 189]},
   {:httpc_handler, :update_session, 4,
    [file: 'httpc_handler.erl', line: 1909]},
   {:httpc_handler, :maybe_make_session_available, 2,
    [file: 'httpc_handler.erl', line: 1516]},
   {:httpc_handler, :answer_request, 3,
    [file: 'httpc_handler.erl', line: 1507]},
   {:httpc_handler, :terminate, 2, [file: 'httpc_handler.erl', line: 759]},
   {:gen_server, :try_terminate, 3, [file: 'gen_server.erl', line: 629]},
   {:gen_server, :terminate, 7, [file: 'gen_server.erl', line: 795]}]}}
   T: :error
   E: :badarg

23:02:32.601 [error] GenServer #PID<0.111.0> terminating
** (stop) {:failed_updating_session, [profile: :httpc_mix, session_id: {{'repo.hex.pm', 443}, #PID<0.111.0>}, pos: 9, value: true, etype: :error, error: :badarg, stacktrace: [{:ets, :update_element, [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.111.0>}, {9, true}], []}, {:httpc_manager, :update_session, 4, [file: 'httpc_manager.erl', line: 210]}, {:httpc_handler, :update_session, 4, [file: 'httpc_handler.erl', line: 1887]}, {:httpc_handler, :maybe_make_session_available, 2, [file: 'httpc_handler.erl', line: 1516]}, {:httpc_handler, :answer_request, 3, [file: 'httpc_handler.erl', line: 1507]}, {:httpc_handler, :terminate, 2, [file: 'httpc_handler.erl', line: 759]}, {:gen_server, :try_terminate, 3, [file: 'gen_server.erl', line: 629]}, {:gen_server, :terminate, 7, [file: 'gen_server.erl', line: 795]}]]}
    (inets) httpc_handler.erl:1911: :httpc_handler.update_session/4
    (inets) httpc_handler.erl:1516: :httpc_handler.maybe_make_session_available/2
    (inets) httpc_handler.erl:1507: :httpc_handler.answer_request/3
    (inets) httpc_handler.erl:759: :httpc_handler.terminate/2
    (stdlib) gen_server.erl:629: :gen_server.try_terminate/3
    (stdlib) gen_server.erl:795: :gen_server.terminate/7
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:ssl, {:sslsocket, {:gen_tcp, #Port<0.4237>, :tls_connection, :undefined}, #PID<0.115.0>}, "HTTP/1.1 200 OK\r\nx-amz-id-2: wtlahJnPAPe2JB3XP009CKKyOlXTjJMC9hPti+9MNEN2wyMmgPHng5nJnNGZjsvO2WNoQfm+xcQ=\r\nx-amz-request-id: 618C205B1E3AC075\r\nx-amz-replication-status: COMPLETED\r\nLast-Modified: Tue, 28 Jun 2016 23:59:12 GMT\r\nETag: \"a5e57ed81098cf777bc319d6118eff50\"\r\nCache-Control: public, max-age=604800\r\nx-amz-meta-surrogate-key: installs\r\nx-amz-version-id: _0bG9Pxir6pB8zBlmt_v1mDuB_StPz9A\r\nContent-Type: text/csv\r\nServer: AmazonS3\r\nVia: 1.1 varnish\r\nFastly-Debug-Digest: 600581fd82da8d7f958c0fc69a76b5f37d0408f6e59492f56cddc3c4a4eb106c\r\nContent-Length: 282\r\nAccept-Ranges: bytes\r\nDate: Wed, 16 Nov 2016 23:02:32 GMT\r\nVia: 1.1 varnish\r\nAge: 452941\r\nConnection: keep-alive\r\nX-Served-By: cache-iad2137-IAD, cache-gru7125-GRU\r\nX-Cache: HIT, HIT\r\nX-Cache-Hits: 2, 70\r\nX-Timer: S1479337352.569482,VS0,VE0\r\n\r\n2.6.2,ff1c5ddfce1fcfd73fd65b8bfc0ff1c13aefc2e98921d528cbc1f35e86c9caa1c9c4e848b9ce6404d9a81c50cfcf0e45dd0dddb23cd42708664c41fce6618900,1.0.0\n2.6.2,ff1c5ddfce1fcfd73fd65b8bfc0ff1c13aefc2e98921d528cbc1f35e86c9caa1c9c4e848b9ce6404d9a81c50cfcf0e45dd0dddb23cd42708664c41fce6618900,1.0.0\n"}
State: {:state, {:request, #Reference<0.0.10.10>, #PID<0.71.0>, 0, :https, {'repo.hex.pm', 443}, '/installs/rebar-1.x.csv', [], :get, {:http_request_h, :undefined, 'keep-alive', :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, 'repo.hex.pm', :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, [], 'Mix/1.3.3', :undefined, :undefined, :undefined, '0', :undefined, :undefined, :undefined, :undefined, :undefined, ...}, {[], []}, {:http_options, 'HTTP/1.1', :infinity, true, {:essl, []}, :undefined, true, :infinity, false}, 'https://repo.hex.pm/installs/rebar-1.x.csv', [], :none, [], 1479337352307, :undefined, :undefined, false}, {:session, {{'repo.hex.pm', 443}, #PID<0.111.0>}, false, :https, {:sslsocket, {:gen_tcp, #Port<0.4237>, :tls_connection, :undefined}, #PID<0.115.0>}, {:essl, []}, 1, :keep_alive, false}, :undefined, :undefined, :undefined, {:httpc_response, :parse, [:nolimit, true]}, {[], []}, {[], []}, :new, [], :nolimit, :nolimit, {:options, {{'proxy.qaservices.locaweb.com.br', 8888}, []}, {{'proxy.qaservices.locaweb.com.br', 8888}, []}, 0, 2, 5, 120000, 2, :disabled, false, :inet, :default, :default, []}, {:timers, [], :undefined}, :httpc_mix, :inactive}
```

**Comment by ericmj** [comment]
@rhruiz That's a bug in OTP but should not cause any issues for us, it only happens when the client is exiting. It has been reported to the OTP team and the next major release should have a fix.

**Comment by simonh1000** [comment]
I had the same issue. As a single data point, I was successful with the Master branch, not least as that version told me that I had dependency resolution error to fix first.

**Comment by ericmj** [comment]
@simonh1000 Thank you for testing and please let us know if the issue comes back when using master.

**Comment by rhruiz** [comment]
@astery @simonh1000 could you describe the network conditions that you get those errors? I do have a large number of moving parts here:  docker, restricted network, http(s) proxy.

**Comment by simonh1000** [comment]
I had an Ethernet connection to a generally good cable broadband

**Comment by astery** [comment]
I able to reproduce this behavior under docker image and out of it. Network is behind nat, internet access is 10 Mbps, no proxy.

Here is some info:
```
$ docker run -it --rm elixir mix -v
Erlang/OTP 19 [erts-8.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

Mix 1.3.4
$ uname -a
Linux dev 4.4.0-47-generic #68-Ubuntu SMP Wed Oct 26 19:39:52 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial
```

**Comment by PotHix** [comment]
I was trying to make it reproducible but it is really hard. I've created the following files:

Dockerfile:
```
FROM elixir:1.3

ADD mix.exs mix.exs
RUN mix do local.hex --force

CMD ["/usr/local/bin/mix", "deps.get"]
```

mix.exs
```
defmodule MixDownload.Mixfile do
  use Mix.Project

  def project do
    [app: :mix_download,
     version: "0.1.0",
     elixir: "~> 1.3",
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     deps: deps()]
  end

  def application do
    [applications: [:logger]]
  end

  defp deps do
    [{:phoenix, "~> 1.2.0"},
     {:postgrex, ">= 0.0.0"},
     {:phoenix_ecto, "~> 3.0"},
     {:phoenix_html, "~> 2.6"},
     {:phoenix_live_reload, "~> 1.0", only: :dev},
     {:phoenix_pubsub, "~> 1.0"},
     {:gettext, "~> 0.9"},
     {:cassette, "~> 1.1"},
     {:cassette_plug, "~> 1.0"},
     {:exrm, "~> 1.0"},
     {:plug, "~> 1.1", override: true},
     {:httpoison, "~> 0.8", override: true},
     {:reverse_proxy, "~> 0.1"},
     {:instream, "~> 0.10"},
     {:elixometer, "~> 1.0"},
     {:exometer_influxdb, github: "travelping/exometer_influxdb"},
     {:exometer_core, "~> 1.4", override: true},
     {:lager, "~> 3.0", override: true},
     {:hackney, "~> 1.4", override: true},
     {:joken, "~> 1.1"},
     {:jose, "~> 1.7"},
     {:calendar, "~> 0.16.0"},
     {:earmark, "~> 1.0", only: [:dev, :test]},
     {:ex_doc, "~> 0.11", only: [:dev, :test]},
     {:credo, "~> 0.3", only: [:dev, :test]},
     {:bypass, "~> 0.1", only: :test},
     {:zookeeper, github: "vishnevskiy/zookeeper-elixir"},
     {:poison, "~> 2.2"},
     {:cowboy, "~> 1.0"}]
  end
end
```

These are dependencies collected from some of our projects. 

So I used my personal computer and one of our servers to run this command for about 30 minutes (20:40 -> 21:10 UTC):

```
while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
```

The result was:

My machine, no proxy
```
.......................** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "calendar"}, 30000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:133: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:123: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4

fail
...........................................................................................................................
```

And our server, using an http proxy (just added two more lines with env vars for `http_proxy` and `https_proxy`)
```
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "cowboy"}, 30000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:133: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:123: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4

fail
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "ex_doc"}, 30000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:133: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:123: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4

fail
.......................................
```

The timeout happens often, but it's really hard to reproduce.

**Comment by ericmj** [comment]
@PotHix Thank you for investigating. Can you reproduce when using the master or emj-disable-pipeline branches?

**Comment by PotHix** [comment]
@ericmj sure. Here are the results:

master branch seems to be worse.

On my machine:
```
$ docker run --rm $(docker build . | grep built | awk '{print $3}') mix hex.info
Hex:    0.14.1-dev
Elixir: 1.3.4
OTP:    19.1.6

Built with: Elixir 1.3.4 and OTP 19.1.6

$ while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
.......Request failed (:timeout)
** (Mix) Package fetch failed and no cached copy available
fail
.........................................................
```

And our server, with proxy is difficult to print here:
```
# docker run --rm $(docker build . | grep built | awk '{print $3}') mix hex.info
Hex:    0.14.1-dev
Elixir: 1.3.4
OTP:    19.1.6

Built with: Elixir 1.3.4 and OTP 19.1.6

# while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "jose"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
..** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "credo"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
...** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "exometer_core"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
```

This is just the beginning. It was 50-50 for fails and successes.

Now with `emj-disable-pipeline` branch:

On my machine:
```
$ cat Dockerfile | grep "mix install"
RUN cd hex && git checkout emj-disable-pipeline && mix install

$ docker run --rm $(docker build . | grep built | awk '{print $3}') mix hex.info
Hex:    0.14.1-dev
Elixir: 1.3.4
OTP:    19.1.6

Built with: Elixir 1.3.4 and OTP 19.1.6

$ while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
..............................................................................................Request failed (:socket_closed_remotely)
** (Mix) Package fetch failed and no cached copy available
fail
..................Request failed (:socket_closed_remotely)
** (Mix) Package fetch failed and no cached copy available
fail
.Request failed (:socket_closed_remotely)
** (Mix) Package fetch failed and no cached copy available
fail
..............
```

On our server, with proxy:
```
# cat Dockerfile | grep "mix install"
RUN cd hex && git checkout emj-disable-pipeline && mix install

# docker run --rm $(docker build . | grep built | awk '{print $3}') mix hex.info
Hex:    0.14.1-dev
Elixir: 1.3.4
OTP:    19.1.6

Built with: Elixir 1.3.4 and OTP 19.1.6

# while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "ex_doc"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
.....** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "gettext"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
....** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "earmark"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
.........** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "cassette_plug"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2
```

**Comment by ericmj** [comment]
I have been able to reproduce by using a proxy. The issue comes from multiple things. Hex sends too many requests too fast so we need to limit the concurrency of requests and have it be configurable by users. This is done here: https://github.com/hexpm/hex/commit/9187bec1dfaaa2c9359d7d9ea537b7e2d5c84906.

On projects with many dependencies this would cause a lot of in-flight concurrent http requests which caused my proxy to reject some. This could also happen without a proxy on networks that are not very good. When the connection errored out when it was in the state of sending the TLS upgrade request it failed to report the error back to the calling process and Hex would just time out without retrying the request. This issue is being fixed here: https://github.com/erlang/otp/pull/1254.

I am not 100% sure if these were the same issues you were seeing and it will be hard for you to test without applying my patches to OTP. But you can try Hex master and if master is still giving you issues try setting a lower `HEX_HTTP_CONCURRENCY`, for example:` HEX_HTTP_CONCURRENCY=1`.

Thanks everyone that has been helping us investigate and reproduce this.

**Comment by astery** [comment]
I tried master branch and it works! It still hangs for while, but after all packages installed without error. Also tried to set `HEX_HTTP_CONCURRENCY=1`, and it works without any timeouts.

**Comment by PotHix** [comment]
@ericmj tried the new master branch and the results are really better! :)

My local machine, no proxy:

```
$ while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
......................................................................................................................................................................................
```

But the default is not good on our server, with proxy:
```
# while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "earmark"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
.** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "ex_doc"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
....** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "httpoison"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
..** (exit) exited in: GenServer.call(#PID<0.112.0>, {:versions, "instream"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:135: Hex.RemoteConverger.check_package_req/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (hex) lib/hex/remote_converger.ex:125: Hex.RemoteConverger.check_input/2
    (hex) lib/hex/remote_converger.ex:41: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:89: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4
    (mix) lib/mix/dep/fetcher.ex:16: Mix.Dep.Fetcher.all/3
    (mix) lib/mix/tasks/deps.get.ex:27: Mix.Tasks.Deps.Get.run/1
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2

fail
```

But the concurrency env var saves the day (`HEX_HTTP_CONCURRENCY=1`):
```
$ while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
.....................................................................................................
```

It also looks OK with `HEX_HTTP_CONCURRENCY=2`:
```
$ while true; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done
......................................
```

Thanks for the patch.

**Comment by ericmj** [comment]
I'm glad master works better and thanks for testing with different values of `HEX_HTTP_CONCURRENCY`. We need to find a good default that works for the majority of people and still has reasonable performance. I will play with different values of it and trying other httpc configurations to see what we can improve.

**Comment by PotHix** [comment]
I will try some values on our server as well. I will post the results here ASAP.

**Comment by PotHix** [comment]
Looks like 20 is a good number.

I used this command with the same Dockerfile and dependencies of the previous test:

```
for i in {11..64}; do echo ""; echo "Trying with HEX_HTTP_CONCURRENCY=$i"; sed -ri "s/HEX_HTTP_CONCURRENCY [0-9]+/HEX_HTTP_CONCURRENCY $i/g" Dockerfile; for i in {1..30}; do printf "."; docker run --rm $(docker build . | grep built | awk '{print $3}') >/dev/null || echo "fail"; done; done >> /root/hex.txt 2>&1
```

It is changing the `HEX_HTTP_CONCURRENCY` environment variable and trying again. 
I've attached the first 30 iterations. It became unusable after that.

[hex.txt](https://github.com/hexpm/hex/files/601944/hex.txt)

**Comment by PotHix** [comment]
@ericmj could you test and define a new number of concurrent downloads?

**Comment by ericmj** [comment]
@PotHix We have already done that, the new default is 8. You can override it by setting `HEX_HTTP_CONCURRENCY`.

**Comment by garmoshka-mo** [comment]
Bug is still there:
(Makes feel really unsure about platform)

````
mix deps.get
Could not find Hex, which is needed to build dependency :amqp
Shall I install Hex? (if running non-interactively, use: "mix local.hex --force") [Yn] Y

12:27:50.838 [error] Failed updating session:
   ProfileName: :httpc_mix
   SessionId:   {{'repo.hex.pm', 443}, #PID<0.119.0>}
   Pos:         9
   Value:       true
when
   Session (db) info: :undefined
   Session (db):      {:session, {{'repo.hex.pm', 443}, #PID<0.119.0>}, false, :https,
 {:sslsocket, {:gen_tcp, #Port<0.4907>, :tls_connection, :undefined},
  #PID<0.120.0>}, {:essl, []}, 1, :keep_alive, false}
   Session (record):  {:EXIT,
 {:badarg,
  [{:ets, :lookup,
    [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.119.0>}], []},
   {:httpc_manager, :lookup_session, 2, [file: 'httpc_manager.erl', line: 189]},
   {:httpc_handler, :update_session, 4,
    [file: 'httpc_handler.erl', line: 1909]},
   {:httpc_handler, :maybe_make_session_available, 2,
    [file: 'httpc_handler.erl', line: 1516]},
   {:httpc_handler, :answer_request, 3,
    [file: 'httpc_handler.erl', line: 1507]},
   {:httpc_handler, :handle_response, 1,
    [file: 'httpc_handler.erl', line: 1251]},
   {:httpc_handler, :handle_info, 2, [file: 'httpc_handler.erl', line: 471]},
   {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]}]}}
   T: :error
   E: :badarg

12:27:50.855 [error] Failed updating session:
   ProfileName: :httpc_mix
   SessionId:   {{'repo.hex.pm', 443}, #PID<0.119.0>}
   Pos:         9
   Value:       true
when
   Session (db) info: :undefined
   Session (db):      {:session, {{'repo.hex.pm', 443}, #PID<0.119.0>}, false, :https,
 {:sslsocket, {:gen_tcp, #Port<0.4907>, :tls_connection, :undefined},
  #PID<0.120.0>}, {:essl, []}, 1, :keep_alive, false}
   Session (record):  {:EXIT,
 {:badarg,
  [{:ets, :lookup,
    [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.119.0>}], []},
   {:httpc_manager, :lookup_session, 2, [file: 'httpc_manager.erl', line: 189]},
   {:httpc_handler, :update_session, 4,
    [file: 'httpc_handler.erl', line: 1909]},
   {:httpc_handler, :maybe_make_session_available, 2,
    [file: 'httpc_handler.erl', line: 1516]},
   {:httpc_handler, :answer_request, 3,
    [file: 'httpc_handler.erl', line: 1507]},
   {:httpc_handler, :terminate, 2, [file: 'httpc_handler.erl', line: 759]},
   {:gen_server, :try_terminate, 3, [file: 'gen_server.erl', line: 629]},
   {:gen_server, :terminate, 7, [file: 'gen_server.erl', line: 795]}]}}
   T: :error
   E: :badarg

12:27:50.880 [error] GenServer #PID<0.119.0> terminating
** (stop) {:failed_updating_session, [profile: :httpc_mix, session_id: {{'repo.hex.pm', 443}, #PID<0.119.0>}, pos: 9, value: true, etype: :error, error: :badarg, stacktrace: [{:ets, :update_element, [:httpc_mix__session_db, {{'repo.hex.pm', 443}, #PID<0.119.0>}, {9, true}], []}, {:httpc_manager, :update_session, 4, [file: 'httpc_manager.erl', line: 210]}, {:httpc_handler, :update_session, 4, [file: 'httpc_handler.erl', line: 1887]}, {:httpc_handler, :maybe_make_session_available, 2, [file: 'httpc_handler.erl', line: 1516]}, {:httpc_handler, :answer_request, 3, [file: 'httpc_handler.erl', line: 1507]}, {:httpc_handler, :terminate, 2, [file: 'httpc_handler.erl', line: 759]}, {:gen_server, :try_terminate, 3, [file: 'gen_server.erl', line: 629]}, {:gen_server, :terminate, 7, [file: 'gen_server.erl', line: 795]}]]}
    (inets) httpc_handler.erl:1911: :httpc_handler.update_session/4
    (inets) httpc_handler.erl:1516: :httpc_handler.maybe_make_session_available/2
    (inets) httpc_handler.erl:1507: :httpc_handler.answer_request/3
    (inets) httpc_handler.erl:759: :httpc_handler.terminate/2
    (stdlib) gen_server.erl:629: :gen_server.try_terminate/3
    (stdlib) gen_server.erl:795: :gen_server.terminate/7
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:ssl, {:sslsocket, {:gen_tcp, #Port<0.4907>, :tls_connection, :undefined}, #PID<0.120.0>}, <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 110, 1, 2, 0, 104, 101, 120, 45, 48, 46, 49, 53, 46, 48, 47, 101, 98, 105, 110, 47, 69, 108, 105, 120, 105, 114, 46, 77, 105, 120, 46, 84, 97, 115, 107, ...>>}
State: {:state, {:request, #Reference<0.0.2.133>, #PID<0.71.0>, 0, :https, {'repo.hex.pm', 443}, '/installs/1.3.0/hex-0.15.0.ez', [], :get, {:http_request_h, :undefined, 'keep-alive', :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, 'repo.hex.pm', :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, :undefined, [], 'Mix/1.3.1', :undefined, :undefined, :undefined, '0', :undefined, :undefined, :undefined, :undefined, :undefined, ...}, {[], []}, {:http_options, 'HTTP/1.1', :infinity, true, {:essl, []}, :undefined, true, :infinity, false}, 'https://repo.hex.pm/installs/1.3.0/hex-0.15.0.ez', [], :none, [], 1487680070263, :undefined, :undefined, false}, {:session, {{'repo.hex.pm', 443}, #PID<0.119.0>}, false, :https, {:sslsocket, {:gen_tcp, #Port<0.4907>, :tls_connection, :undefined}, #PID<0.120.0>}, {:essl, []}, 1, :keep_alive, false}, {'HTTP/1.1', 200, 'OK'}, {:http_response_h, 'public, max-age=604800', 'keep-alive', 'Tue, 21 Feb 2017 12:27:50 GMT', :undefined, :undefined, :undefined, :undefined, '1.1 varnish', :undefined, 'bytes', '354703', '"32639417cd22865fe37dd2eba7597f8a"', :undefined, :undefined, :undefined, 'AmazonS3', :undefined, :undefined, :undefined, :undefined, :undefined, '478371', :undefined, :undefined, :undefined, 'application/andrew-inset', :undefined, 'Fri, 23 Dec 2016 23:29:31 GMT', [{'x-amz-id-2', 'YcuzbLY69wJibQhxHS7Ab+vWP+pTax+yGM1tOYEYOuEp/rOmeUrWjnMU+NtwdILC9M/qM+uCaDQ='}, {'x-amz-request-id', '55D20D92FB98DA4B'}, {'x-amz-replication-status', 'FAILED'}, {'x-amz-meta-surrogate-key', 'installs'}, {'x-amz-version-id', 'OClpliznSKNcQv3EnXaKQigPoUrT5Ftm'}, {'fastly-debug-digest', '5db161298266899fd3e93f292ed51788ae6e376328c4d76c0190675b35234c98'}, {'x-served-by', 'cache-iad2130-IAD, cache-hhn1534-HHN'}, {'x-cache', 'HIT, HIT'}, {'x-cache-hits', '1, 1'}, {'x-timer', 'S1487680070.414301,VS0,VE0'}]}, :undefined, {:httpc_response, :whole_body, [<<80, 75, 3, 4, 20, 0, 0, 0, 0, 0, 137, 3, 152, 73, 173, 155, 214, 230, 32, 108, 0, 0, 32, 108, 0, 0, 34, 0, 0, 0, 104, 101, 120, 45, 48, 46, 49, 53, 46, 48, ...>>, 478371]}, {[], []}, {[], []}, :new, [], :nolimit, :nolimit, {:options, {:undefined, []}, {:undefined, []}, 0, 2, 5, 120000, 2, :disabled, false, :inet, :default, :default, []}, {:timers, [], :undefined}, :httpc_mix, :inactive}
````

**Comment by josevalim** [comment]
> (Makes feel really unsure about platform)

To be fairly honest, I would feel the same. However, I want to clarify that httpc, which is the http client that ships as part of Erlang/OTP, does not reflect the platform as a whole. We are aware of the limitations of the built-in http client and we are working on addressing them.

For anyone wanting to use an HTTP client in your applications, we recommend [hackney](https://github.com/benoitc/hackney). The only reason Hex cannot use hackney is because we need something built-in in order to bootstrap the package manager (i.e. we need to have something that will download hackney to your system).

Did you try again and did it work the second time?

**Comment by garmoshka-mo** [comment]
@josevalim Yes, after few attempts (it shown same error on different dependencies), package manager finally did his job and I was able to execute required library. 

Just pointing on emotions, which feel user, who seeing such behaviour at the very first steps. That could make him concerned about going further - can be significant barrier for community growth. (Maybe it's pretty important issue.)

**Comment by ericmj** [comment]
@garmoshka-mo You are talking about a separate bug which is only a warning - not an error. Please include the output of `mix hex.info` when reporting errors. This warning should be fixed in OTP 19.2.

**Comment by theocodes** [comment]
Same here :(

I was getting:

```
Failed to fetch record for 'ecto' from registry (using cache)
:socket_closed_remotely
```

until I ran:

`mix deps.clean --all`

and now getting:

```
[user]mix deps.get                                                                                                                                                                                                                                               ruby-2.0.0-p353 
** (exit) exited in: GenServer.call(#PID<0.121.0>, {:versions, "comeonin"}, 60000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:737: GenServer.call/3
    (hex) lib/hex/remote_converger.ex:208: Hex.RemoteConverger.check_dep/2
    (elixir) lib/enum.ex:651: anonymous fn/3 in Enum.each/2
    (stdlib) lists.erl:1263: :lists.foldl/3
    (elixir) lib/enum.ex:1772: Enum.each/2
    (hex) lib/hex/remote_converger.ex:38: Hex.RemoteConverger.converge/2
    (mix) lib/mix/dep/converger.ex:91: Mix.Dep.Converger.all/4
    (mix) lib/mix/dep/converger.ex:50: Mix.Dep.Converger.converge/4

[user]mix hex.info                                                                                                                                                                                                                                               ruby-2.0.0-p353 
Hex:    0.15.0
Elixir: 1.4.1
OTP:    19.2.3

Built with: Elixir 1.3.4 and OTP 18.3.4.4
```

**Comment by josevalim** [comment]
Some folks are reporting Amazon S3 is returning 500 and timeouts for them. It seems to be an S3 issue (which is where packages are hosted).

**Comment by josevalim** [comment]
More info here: https://status.aws.amazon.com

**Comment by MonkeyIsNull** [comment]
Thanks for the update, I just got clobbered by this problem.

**Comment by theocodes** [comment]
Thanks @josevalim! Was beginning to pull my hair out! Oh wait, i have none left :(
